random-distribution-helpers.go

```diff
@@ -82,43 +88,53 @@ func (distSampler RandomDistributionSampler) Set(value string) error {
 		if len(params) != 1 {
 			return fmt.Errorf(formatErr, "Constant distribution expects exactly one parameter but got %v.", len(params))
 		}
-		if value, err := time.ParseDuration(typeAndParams[1]);  err == nil {
-			distSampler.distribution = RandomConstDistribution{value: value}
+		if value, err := distSampler.parseDuration(typeAndParams[1]); err == nil {
+			distSampler.distribution = &RandomConstDistribution{value: value}
 		} else {
 			return fmt.Errorf(formatErr, err)
 		}
 	case "equal": // Parse values for equal distribution
 		if len(params) != 2 {
 			return fmt.Errorf(formatErr, "Equal distribution expects exactly two parameters but got %v.", len(params))
 		} else {
-			min, err := time.ParseDuration(params[0])
+			min, err := distSampler.parseDuration(params[0])
 			if err != nil {
 				return fmt.Errorf(formatErr, err)
 			}
 			max, err := time.ParseDuration(params[1])
 			if err != nil {
 				return fmt.Errorf(formatErr, err)
 			}
-			distSampler.distribution = RandomEqualDistribution{min: min, max: max}
+			distSampler.distribution = &RandomEqualDistribution{min: min, max: max}
 		}
 	case "norm": // Parse values for normal distribution
 		if len(params) != 2 {
 			return fmt.Errorf(formatErr, "Normal distribution expects exactly two parameters but got %v.", len(params))
 		} else {
-			mu, err := time.ParseDuration(params[0])
+			mu, err := distSampler.parseDuration(params[0])
 			if err != nil {
 				return fmt.Errorf(formatErr, err)
 			}
-			sigma, err := time.ParseDuration(params[1])
+			sigma, err := distSampler.parseDuration(params[1])
 			if err != nil {
 				return fmt.Errorf(formatErr, err)
 			}
-			distSampler.distribution = RandomNormalDistribution{mu: mu, sigma: sigma}
+			distSampler.distribution = &RandomNormalDistribution{mu: mu, sigma: sigma}
 		}
 	default:
 		return fmt.Errorf(formatErr, "Unknown distribution type identifier %v.", typeAndParams[0])
 	}
 	log.Printf("Successfully parsed distribution parameter %v. Result: %v", value, distSampler.distribution.String())
 
 	return nil
-}
\ No newline at end of file
+}
+
+func (distSampler RandomDistributionSampler) parseDuration(value string) (time.Duration, error) {
+	timeValue, err := time.ParseDuration(value)
+	if err == nil && timeValue >= 0 {
+		return timeValue, nil
+	} else if err == nil && timeValue < 0 {
+		return -1, fmt.Errorf("Distribution argument must be a positive time value but actually is %v", timeValue)
+	}
+	return -1, err
```

Comment:
The if conditions in this method can be formulated in a way that the -1 return value (which cannot be reached) is not necessary. First check for the error, and return if it is non-nil, then check if timeValue is above or below zero.
Also, Since this method does not use the receiver distSampler, it should be turned into a function.

Reply:
I re-adjusted the order of return values. However, both -1 value can be reached. First, the format can be invalid, thus the parsing fails. Second, the time value must not be negative (the Duration-parsing allows negative values). This requires a different error message imo. Third, everything is fine.